src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs

```diff
@@ -482,44 +482,50 @@ public static bool TryValidateBlock(BinaryReader reader, ZipArchiveEntry entry)
             }
             else
             {
-                if (reader.BaseStream.Length < reader.BaseStream.Position + extraFieldLength + entry.CompressedLength + 4)
+                if (reader.BaseStream.Length < reader.BaseStream.Position + extraFieldLength + entry.CompressedLength + 12)
                 {
                     return false;
                 }
 
                 reader.BaseStream.Seek(extraFieldLength + entry.CompressedLength, SeekOrigin.Current); // seek to end of compressed file from which Data descriptor starts
-                uint dataDescriptorSignature = reader.ReadUInt32();
-                bool wasDataDescriptorSignatureRead = false;
-                int seekSize = 0;
-                if (dataDescriptorSignature == DataDescriptorSignature)
+                Span<uint> buffer = stackalloc uint[6];
+
+                // let's try to match the smallest possible structure (3 x 4) 32 bit without signature
+                buffer[0] = reader.ReadUInt32();
+                buffer[1] = reader.ReadUInt32();
+                buffer[2] = reader.ReadUInt32();
+
+                if (TestMatch(entry, DataDescriptorSignature, buffer[0], buffer[1], buffer[2]))
                 {
-                    wasDataDescriptorSignatureRead = true;
-                    seekSize = 4;
+                    return true;
                 }
 
-                bool is64bit = entry._versionToExtract >= ZipVersionNeededValues.Zip64;
-                seekSize += (is64bit ? 8 : 4) * 2;   // if Zip64 read by 8 bytes else 4 bytes 2 times (compressed and uncompressed size)
+                // let's try to match the next record size (4 x 4) 32 bit with signature
+                buffer[3] = reader.ReadUInt32();
```

Comment:
Do we need to check if we can read the next `uint`? Above we check if we can read the first 12 bytes. But here are we guaranteed that there will be more to read?

Reply:
Not guaranteed, but expected, sure will add check